modules/tts/tts-adapters/tts-adapter-sapinative/src/main/java/org/daisy/pipeline/tts/sapi/impl/SAPIEngine.java

```diff
@@ -93,33 +88,55 @@ public AudioInputStream speak(String ssml, Voice voice, TTSResource resource, Li
 		voice = mVoiceFormatConverter.get(voice.name.toLowerCase());
 		ThreadResource tr = (ThreadResource)resource;
 		if (voice.engine.equals("sapi") ){
-			int res = SAPI.speak(tr.SAPIConnection, voice.engine, voice.name, ssml);
-			if (res != SAPIResult.SAPINATIVE_OK.value()) {
-				throw new SynthesisException("SAPI-legacy speak error " + res + " raised with voice "
-				                             + voice + ": " +  SAPIResult.valueOfCode(res)+"\nFor text :"
-				                             + ssml);
+			try{
+				int res = SAPI.speak(tr.SAPIConnection, voice.engine, voice.name, ssml);
+				if (res != SAPIResult.SAPINATIVE_OK.value()) {
+					throw new SynthesisException("SAPI-legacy speak error " + res + " raised with voice "
+							+ voice + ": " +  SAPIResult.valueOfCode(res)+"\nFor text :"
+							+ ssml);
+				}
+			} catch (RuntimeException e){
+				Logger.error("SAPI-legacy raised a RUNTIME exception while speaking " + ssml + " with " + voice + " : " + e.getMessage());
+				throw new SynthesisException("SAPI-legacy raised a RUNTIME exception while speaking " + ssml + " with " + voice, e);
+			} catch (Exception e){
+				Logger.error("SAPI-legacy raised an exception while speaking " + ssml + " with " + voice + " : " + e.getMessage());
+				throw new SynthesisException("SAPI-legacy raised an exception while speaking " + ssml + " with " + voice, e);
 			}
+
 			int size = SAPI.getStreamSize(tr.SAPIConnection);
 			byte[] data = new byte[size];
 			SAPI.readStream(tr.SAPIConnection, data, 0);
-			long[] bookmarksPositions = SAPI.getBookmarkPositions(tr.SAPIConnection);
+
+			String[] names = SAPI.getBookmarkNames(tr.SAPIConnection);
+			long[] positions = SAPI.getBookmarkPositions(tr.SAPIConnection);
 			float sampleRate = sapiAudioFormat.getSampleRate();
 			int bytesPerSample = sapiAudioFormat.getSampleSizeInBits() / 8;
-			for (long position : bookmarksPositions) {
-				int offset = (int) ((position * sampleRate * bytesPerSample) / 1000);
-				marks.add(offset);
+			for(int i = 0; i < names.length; ++i){
+				int offset = (int) ((positions[i] * sampleRate * bytesPerSample) / 1000);
+				// NP : it happens that SAPI / OneCore sometimes make empty bookmarks
+				// (i have not found why yet)
```

Comment:
@NPavie: are they extra bookmarks that don't correspond to a `mark` element in the SSML? It is important that the [contract of `TTSEngine`](http://daisy.github.io/pipeline/api/org/daisy/pipeline/tts/TTSEngine.SynthesisResult.html#marks) remains satisfied.

Reply:
> are they extra bookmarks that don't correspond to a mark element in the SSML? 

Indeed : on some occasions in my test, i was having registered "bookmark" events with no duration nor name assigned, and that were not linked to a mark in the ssml (all marks were also correctly registered and reported by the connector with their name and duration).

This was causing error in the process where the pipeline was raising an error stating there were more marks than excepted.

Reply:
OK thanks